common: implement random number generator (following N3551)

chardan opened the pull request:
Adds new mini-library for random number generation using C++11's
`<random>` header.

Although a useful simplified interface into `<random>` on its own, this is also a dependency for PRs based on the following set of changes:
https://github.com/chardan/ceph/tree/jfw-wip-puppies_against_rand

The implementation is (considerably) extended from the simplified `<random>` interface discussed in:
https://isocpp.org/files/papers/n3551.pdf

In addition to providing improved-quality random number generation, this facility also adds locking, toward fixing:
http://tracker.ceph.com/issues/18873

By default, the mini-library provides internal locks and global random number generators, but also provides reasonably lightweight mechanisms for users to hook in their own locks and generators. (More complex uses are out of the scope of this mini-library and left to use the full power of the `<random>` facility, or implementations of their choosing.)

Signed-off-by: Jesse Williamson <jwilliamson@suse.de>

smithfarm commented on src/include/random.h:
```diff
@@ -0,0 +1,113 @@
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2012 Inktank Storage, Inc.
+ * Copyright (C) 2014 Red Hat <contact@redhat.com>
```
Are we sure this copyright notice is correct? Isn't this new code?

chardan replied:
Thank you for noticing! What is the right way to introduce a new module? Who gets the copyright?

tchaikov replied:
IANAL, in most cases, it belongs to your employer if you are not doing this in your spare time and using your own devices.

chardan replied:
Ok, I'll take care of it. Thanks!

smithfarm commented on src/test/common/test_random.cc:
```diff
@@ -0,0 +1,66 @@
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2011 New Dream Network
```
Check copyright notice

chardan commented on src/include/random.h:
```diff
@@ -0,0 +1,113 @@
+ detail::urng<EngineT>().seed(seed);
+}
+
+template <int max = std::numeric_limits<int>::max(), int min = 0, 
```
This should also use std::numeric_limits<>::min().

chardan commented on src/include/random.h:
```diff
@@ -0,0 +1,113 @@
+int generate_random_number(const IntegerT max,
+                           typename std::enable_if<std::is_integral<IntegerT>::value>::type* = 0)
+{
+ constexpr decltype(max) zero = 0;
```
Alternatively, IntegerT seems fine.

chardan commented on src/include/random.h:
```diff
@@ -0,0 +1,113 @@
+RealT generate_random_number(const RealT max,
+                             typename std::enable_if<std::is_floating_point<RealT>::value>::type* = 0)
+{
+ constexpr decltype(max) zero = 0.0;                  // no variable templates in C++11 :-(
```
(Same as above, we already have RealT.)

chardan commented on src/test/common/test_random.cc:
```diff
@@ -0,0 +1,66 @@
+ auto a = ceph::util::generate_random_number();
+ auto b = ceph::util::generate_random_number();
+
+ // Technically, this can still collide and cause a false test failure, but let's be optimistic:
```
How do people feel about this? This test may cause occasional false-positives.

chardan commented on src/test/common/test_random.cc:
```diff
@@ -0,0 +1,66 @@
+ auto b = ceph::util::generate_random_number(std::numeric_limits<int>::max(), 1);
+
+ if(std::numeric_limits<int>::max() > 32767)
+  ASSERT_NE(a, b);
```
An additional check might be for not-zero. Maybe it would be more interesting to narrow the range? I don't want this testing the RNG itself.

chardan commented on src/test/common/test_random.cc:
```diff
@@ -0,0 +1,66 @@
+ }
+
+ {
+ auto a = ceph::util::generate_random_number<6>();
```
I should do some less lazy due diligence: not gt 6, not lt 0...

tchaikov commented:
@chardan do we have any user of this rng implementation now?

chardan commented:
@tchaikov Not at present. 

In terms of accompanying code WRT Ceph, this is a dependency for PRs based on the following set of changes:
https://github.com/ceph/ceph/compare/master...chardan:jfw-wip-puppies_against_rand

The implementation is extended from the post referenced at:
https://isocpp.org/files/papers/n3551.pdf

lumneh commented:
 X 

The same confnicSent from my Samsung Galaxy smartphone.
-------- Original message --------From: Jesse Williamsonctcc <notifications@github.com> Date: 29/05/2017  18:22  (GMT+04:00) To: ceph/ceph <ceph@noreply.github.com> Cc: Subscribed <subscribed@noreply.github.com> Subject: Re: [ceph/ceph] common: implement random number generator (following N3551) (#15341) 
@tchaikov Not at present.
In terms of accompanying code WRT Ceph, this is a dependency for PRs based on the following set of changes:

master...chardan:jfw-wip-puppies_against_rand
The implementation is extended from the post referenced at:

https://isocpp.org/files/papers/n3551.pdf

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/ceph/ceph","title":"ceph/ceph","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/ceph/ceph"}},"updates":{"snippets":[{"icon":"PERSON","message":"@chardan in #15341: @tchaikov Not at present. \r\n\r\nIn terms of accompanying code WRT Ceph, this is a dependency for PRs based on the following set of changes:\r\nhttps://github.com/ceph/ceph/compare/master...chardan:jfw-wip-puppies_against_rand\r\n\r\nThe implementation is extended from the post referenced at:\r\nhttps://isocpp.org/files/papers/n3551.pdf\r\n\r\n "}],"action":{"name":"View Pull Request","url":"https://github.com/ceph/ceph/pull/15341#issuecomment-304672468"}}}

lumneh commented:
Sent from my Samsung Galaxy smartphone.
-------- Original message --------From: Jesse Williamson <notifications@github.com> Date: 29/05/2017  18:22  (GMT+04:00) To: ceph/ceph <ceph@noreply.github.com> Cc: Subscribed <subscribed@noreply.github.com> Subject: Re: [ceph/ceph] common: implement random number generator (following N3551) (#15341) 
@tchaikov Not at present.
In terms of accompanying code WRT Ceph, this is a dependency for PRs based on the following set of changes:

master...chardan:jfw-wip-puppies_against_rand
The implementation is extended from the post referenced at:

https://isocpp.org/files/papers/n3551.pdf

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/ceph/ceph","title":"ceph/ceph","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/ceph/ceph"}},"updates":{"snippets":[{"icon":"PERSON","message":"@chardan in #15341: @tchaikov Not at present. \r\n\r\nIn terms of accompanying code WRT Ceph, this is a dependency for PRs based on the following set of changes:\r\nhttps://github.com/ceph/ceph/compare/master...chardan:jfw-wip-puppies_against_rand\r\n\r\nThe implementation is extended from the post referenced at:\r\nhttps://isocpp.org/files/papers/n3551.pdf\r\n\r\n "}],"action":{"name":"View Pull Request","url":"https://github.com/ceph/ceph/pull/15341#issuecomment-304672468"}}}

chardan commented:
@smithfarm Do those look in-order now?

cbodley commented on src/include/random.h:
```diff
@@ -0,0 +1,115 @@
+{
+ static T rng;
+ return rng;
+}
```
note that a static engine is not thread-safe. `Fixes: http://tracker.ceph.com/issues/18873` seems to imply otherwise

to be thread-safe, applications either need to provide mutual exclusion when accessing a shared engine, or guarantee that a given engine is only accessed from a single thread

chardan replied:
Yes, that's correct. The feature was not originally implemented with issue 18873 in mind, but now that it's changed, I'll address it. (See further comments.)

chardan commented on src/include/random.h:
```diff
@@ -0,0 +1,115 @@
+          typename EngineT>
+NumberT generate_random_number(const NumberT max, const NumberT min)
+{
+ static DistributionT d { min, max };
```
Right, but originally this was a feature: the idea is to not have to have extraneous RNGs lying around, instead one-per-type. I'll revisit, though I still like this concept.

cbodley replied:
the distribution itself is trivial to construct, it just calls the param_type ctor

chardan commented on src/include/random.h:
```diff
@@ -0,0 +1,115 @@
+
+template <typename IntegerT>
+IntegerT generate_random_number(const IntegerT max, const IntegerT min,
+                                typename std::enable_if<std::is_integral<IntegerT>::value>::type* = 0)
```
"Trying to be too clever." I have waffled back and forth on this, originally liking that it was consistent with the single-parameter version... however, I think changing it to (min, max) is a muh better interface. Fortunately, I'll now have the opportunity to fix it! Thank you for commenting.

chardan commented:
I am going to give this some extra TLC, as it appears the scope has expanded from the original feature in light of issue 18873. Thank you for your comments!

chardan commented:
@cbodley, this is considerably reworked. If folks are happy with the way this looks (see test file for canonical use example) and works, the next step is a collection of PRs migrating calls to rand()/srand() to this mechanism. Thanks for taking a look!

adamemerson commented on src/include/random.h:
```diff
@@ -0,0 +1,391 @@
+ std::lock_guard<MutexT> lg(m);
+ detail::urng<EngineT>().seed(n);
+}
+
```
Being able to reseed a GLOBAL random number source like this seems dubious. There seems a risk of some portion of the program using a known, determined seed and leaving it there when other parts want unpredictable random data.

adamemerson replied:
If we used thread_local random variables doing it scoped might be better.

chardan replied:
I like your suggestion of thread_local, and will see what I can do. Thanks!

chardan replied:
Also note that the library provides a facility for the user to specify their own rng or get a unique one (see test file). Open question: I don't really like encouraging folks to use unit tests as example code, is there a place for us to put internal examples?

adamemerson replied:
How do you feel about something like...

https://gist.github.com/adamemerson/ca2a8fa70f2fc592740469372ee83026

?

chardan replied:
Hi! Thanks for your patience while I got the gist of this-- here are some non-random thoughts. 
It looks like the purpose of this is to improve the quality of the random number seeding. I think that might be a good idea, but also that it should come as a follow up PR. 

The change at hand is meant to provide an alternative to rand(), srand(), and
mod, while hiding some of the details of <random> from the user (aka "easy
to use"). The default random engine is "supposedly" good enough for this  
purpose, given that here there's no attempt to generate cryptograpically 
strong random numbers.

I'm definitely aware that there are some criticisms of several aspects of
<random>, and another hopeful side-effect is that when something like     
N4316
("http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4316.html") is
adopted, we will easily be able to change to using it with everything
being centralized. (As for deeper criticism of <random>, I'm going to stay
out of that one-- it might be worthwhile to send proposals to the
committee, or talk with someone involved.)

It might even be possible to hook seeding improvements like your suggestion shows into this randomization
mini-library via the engine hooks in this PR, which lets the
user choose and would avoid affect any code. (Assuming it's not best to nip in the bud now.)

Ultimately, this is a nerfed version of <random>: if we
get /too/ general, or the user has a very specific or expert-domain
problem, they probably should use <random> directly, or the tools of their
choice. I'm also ok with replacing the RNG seeding if folks feel strongly
about it and think it's going to cover the cases needed. My goal here is  
to do better than rand() and deliver thread safety and /some/ flexibility.

I wouldn't want to see the baby thrown out with the bathwater-- certainly,
since it sounds like there's utility in providing stronger seeding it
should be considered-- I think a PR based on your extension would be valuable.

adamemerson replied:
That seems reasonable. In that case my recommendation would be to take the 'nerfed' version of <random> and rework it to use all thread_local variables instead of locks.

Regarding seeding, I would say simply don't have a way to 'seed' the nerfed global random engines, and if someone WANTS to seed something, make them use the random_number_generator class that you have. That way we avoid the risk of some subsystem seeding a global/thread_local RNG with a deterministic value and forgetting to re-randomize it and reset all the distributions it used.

chardan replied:
I am a bit concerned about the idea of relying only on thread_local: if an engine has been provided that could be externally be pointed to across threads, we could still have contention on the engine. One idea: we can say that if the user does /not/ pass a mutex (or use, perhaps, a "locked" class) then things are thread_local. If the user /does/ pass a mutex, we lock. Doesn't seem too bad to me.

@adamemerson Let me think a while more about seeding. Your suggestion sounds pretty good, but maybe we need to leave a single randomize_rng() in-- the one with no parameters (or mutex), which will internally just use std::random_device. Then again, I do like the idea that using one of the higher-level constructs (the function generator or function object) implies a unique engine and thus avoids this situation entirely. I'll ponder a bit.

@cbodley thoughts?

chardan commented:
Two more commits added; one removes locks and relies on thread_local storage when functions without a lock are called; the other removes any explicit re-seeding (the user may still call randomize_rng()).
@cbodley @adamemerson

chardan commented:
Jenkins retest this please.

adamemerson commented:
I think that sounds fine! A forced 'reseed' that isn't deterministic is harmless. I don't think anyone will, in practice, USE, the mutectacular versions, but there's no harm in having them available.

chardan commented:
Do the tests look adequate? Anything else I should be checking?

adamemerson commented:
Given that we're wrapping someone else's RNG (and testing RNGs is /awfully difficult/) all we really need is Sanity and Reasonableness and I think the tests do that.

A probabilistic test like none_equal worries me a /tiny/ bit since I really hate false failure reports, but I don't expect it to actually fire in PRACTICE. If there's a simple way we can do without it, it might be better, but probably not necessary.

chardan commented:
@adamemerson Agreed, I think I even made a comment about that in the source. OTOH, I had trouble thinking of a decent improvement. If it /does/ trigger, it will be pretty annoying because of the long build/test time. I can also take them out, though I don't mind at least checking. Testing the RNG itself sounds like a job for the standard library implementers and specialists-- my own knowledge of the topic isn't that deep.

chardan commented:
@tchaikov @joscollin @cbodley

chardan commented:
@tchaikov Hopefully these changes move everything closer to the right direction. If you have specific concerns, please let me know.

tchaikov commented on src/include/random.h:
```diff
@@ -0,0 +1,398 @@
+template <typename T>
+T& engine()
+{
+ thread_local T rng_engine;
```
https://github.com/ceph/ceph/blob/master/CodingStyle#L76
```md
* Formatting > Tabs:
  Indent width is two spaces.  When runs of 8 spaces can be compressed
  to a single tab character, do so. The standard Emacs/Vim settings
  header is:
```

tchaikov replied:
@chardan please respect the coding style. indent with two spaces.

chardan commented:
Jenkins retest this please.

chardan commented:
Jenkins retest this please.

joscollin commented:
@chardan Sorry for coming late, as I was busy with other tasks.

@chardan @badone :
I hope this PR eventually addresses the following Coverity Scan report:

```
*** CID 1412976:  Security best practices violations  (DC.WEAK_CRYPTO)
/home/brad/working/src/ceph/src/test/librbd/test_librbd.cc: 5530 in TestLibRBD_ExclusiveLock_Test::TestBody()::[lambda(int, void *&) (instance 1)]::operator ()(int, void *&) const()
5524     	EXPECT_TRUE(lock_owner);
5525     	std::cout << m_id << ": exclusive lock acquired" << std::endl;
5526     	{
5527     	  lock_guard<mutex> locker(lock);
5528     	  owner_id = m_id;
5529     	}
>>>     CID 1412976:  Security best practices violations  (DC.WEAK_CRYPTO)
>>>     "rand" should not be used for security related applications, as linear congruential algorithms are too easy to break.
5530     	usleep(rand() % 50000);
5531           }
5532     
5533           lock_guard<mutex> locker(lock);
5534           if (owner_id == m_id) {
5535     	EXPECT_EQ(0, rbd_lock_release(m_image));

** CID 1412977:  Security best practices violations  (DC.WEAK_CRYPTO)
/home/brad/working/src/ceph/src/os/bluestore/BlueStore.cc: 2027 in BlueStore::ExtentMap::allocate_spanning_blob_id()()
```

adamemerson commented:
That warning seems..odd.

On the one hand we shouldn't be using rand at all because it's not thread-safe.

On the other hand, for anything that IS actually security critical we should probably get an RNG from our crypto library since they've been audited well, and not just use the one from the C++ standard library.

chardan commented:
Jenkins retest this please.

chardan commented:
@joscollin I have a series of PRs more-or-less ready that depend on this one... and, I believe they will address rand() calls like that one.
@adamemerson Right, the goal of this feature is to provide something better than rand(). For specialist tasks, we'll still need specialist attention. This library does provide facilities for using any engine compatible with the standard library, and the default one still has a larger period than rand()'s, so in addition to thread safety I think there will be improvement in that direction.

Thoughts from everyone? Are we looking ok now?

chardan commented:
@adamemerson Another thought: maybe the static analyzer is simply checking for that kind of use of rand(), and isn't really all that aware of the context?

@tchaikov Whitespace has been revised, especially leading WS and spaces before control structures,--
 though I'm a bit unclear on the tab rules and in some situations (like function parameter alignment) it doesn't really seem like it's an improvement. Are things more-or-less where you'd like them to be?

If things are looking good to most everyone, I'm looking forward to moving to the next non-random step! Thanks.

tchaikov commented on src/test/common/CMakeLists.txt:
```diff
@@ -88,6 +88,13 @@ add_executable(unittest_util
 add_ceph_unittest(unittest_util ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/unittest_util)
 target_link_libraries(unittest_util global ${BLKID_LIBRARIES})
 
+# unittest_random
```
remove the comment, if you please.

chardan replied:
I'm happy to (done), but had added it because that seems to be the general pattern in the existing file.

tchaikov commented on src/test/common/test_random.cc:
```diff
@@ -0,0 +1,274 @@
+TEST(util, test_make_random_function)
+{
+  // Test convenience functions:
+ 
```
please remove the empty line.

tchaikov commented on src/test/common/test_random.cc:
```diff
@@ -0,0 +1,274 @@
+  constexpr auto max_int   = std::numeric_limits<int>::max();
+ 
+  {
+  auto rng_i = ceph::util::make_random_number_function<1, max_int - 1>();
```
wrong indent.

chardan replied:
Ok, to clarify: non-control blocks are to be spaced as-per control blocks?

tchaikov commented on src/test/common/test_random.cc:
```diff
@@ -0,0 +1,274 @@
+ 
+  {
+  auto a = rng(10);
+  ASSERT_LT(a, 11);
```
```c++
ASSERT_LE(a, 10);
```

chardan replied:
Oh! Derp. Thanks.

tchaikov commented on src/test/common/test_random.cc:
```diff
@@ -0,0 +1,274 @@
+  }
+ 
+  {
+  auto a = rng(10.0);
```
what's the point of this test? to test that 10.0 can be casted to `int`?

chardan replied:
In this case, I'd expect variable 'a' to be a float. It's just checking that the mechanism compiles. The value should be between 0.0 and 10.0.

tchaikov replied:
no, `rng` is an integer random number generator, and `a` will be an integer. this test is not verifying your expectation.

tchaikov replied:
@chardan ping?

chardan commented on src/include/random.h:
```diff
@@ -0,0 +1,397 @@
+}
+} // namespace detail
+
+namespace detail {
```
I see them as being generally unrelated to each other, there's no technical reason.

chardan commented:
@tchaikov Thank you for your feedback! Most of your comments should be addressed, thanks in particular for pointing out the GTE/LTE macros!

chardan commented:
Hi everyone! Where are we at, here?

chardan commented:
Hi folks, how are things looking here? Thank you!
@tchaikov @adamemerson @joscollin @cbodley

adamemerson commented:
It should be good.

tchaikov commented on src/test/common/test_random.cc:
```diff
@@ -0,0 +1,272 @@
+  // min > max should not explode:
+  {
+    float min = 1.0, max = 0.0;
+    auto x = ceph::util::generate_random_number(min, max);
```
```
warning: unused variable ‘x’ [-Wunused-variable]
```

tchaikov commented on src/test/common/test_random.cc:
```diff
@@ -0,0 +1,272 @@
+  gen_fn.seed(42);   // re-seed
+ 
+  // Mostly to swallow "unused variable" warnings:
+  ASSERT_EQ(true, none_equal(a, b, c, d, e, x, y, z));
```
could use `ASSERT_TRUE` instead.

cbodley commented on src/test/common/test_random.cc:
```diff
@@ -0,0 +1,272 @@
+ 
+  ceph::util::randomize_rng(l);
+ 
+  ceph::util::generate_random_number(l);
```
i don't like how this part of the interface mixes global engines with caller-provided locks. this is error-prone, because it relies on the caller to provide the same lock for every call in order for it to actually be thread-safe. a global engine should be protected by a global mutex

adamemerson commented:
We could just rely on the thread-local engines and kill the lock-protected interface entirely. I think if the user wants something more specific they can make that, but thread-local should give us everything we need in a simple version.

cbodley commented:
> We could just rely on the thread-local engines and kill the lock-protected interface entirely

this sounds better, yeah - but how can we guarantee that each thread-local engine gets seeded correctly?

adamemerson commented:
I thought there was something to handle that but I'm not seeing it in a quick glance.

I have an example in https://gist.github.com/adamemerson/ca2a8fa70f2fc592740469372ee83026 for handling per-thread RNG initialization on first use.

adamemerson commented:
Ignore the seed-guard stuff, I was just thinking of a way to 'safely' reseed for deterministic use that didn't risk having outside code effective. But I'm sort of convinced that if a function actually needs that other than just 'good source of random bits' they should use the <random> functions directly.

chardan commented:
Thanks everyone for the constructive feedback! I think we're getting close here.

I. As currently designed, the ability for users to specify their own locks and engines is
a feature of the library. Doing so is simpler than interacting directly with <random>.
OTOH, for example, there's not a way to specify a unique mutex to the function object
interface. OTOOH, using either an explicit lock or engine requires explicit syntax for 
doing so, so I think the potential for misuse is relatively low-- less than when using
the standard library where any lock is potentially away from the call site (if this is
unclear, please ask and I'll show what I mean in code).

Proposed options:
1) leave it as-is (my vote, TBH-- I think the balance is about right, simple and safe normal behavior,
a bit of an escape hatch for more advanced things, but not too overreaching);
2) make the current lock/engine interface an implementation detail and wait-and-see if
there's any actual need to expose it (the implementation uses it anyway, but if the consensus is
that it's dangerous to expose, we don't have to);

II. WRT seeding:

This isn't so easy. Right now, the requirement is to having the application seed the RNG 
at startup (and it may be re-seeded safely later), and my plan is to patch in exactly that. 
However, the function object interface may also be used (it will be seeded on construction, 
and may be re-seeded), and if you need more granularity you could build your own on top 
of the existing support without too much further strain (because it allows custom locks and engines).
(This situation is no worse than and even improved from rand() and srand().)

adamemerson commented:
So, the thing I've never been clear on (I didn't insist we NOT do it because I assume you have a reason for wanting it) is, what exactly is the ADVANTAGE of having the mutex as an optional parameter at all when we have the thread_local version of the interface?

I don't really see any reason for the mutext to be desirable in a simple library like this one.

chardan commented:
@adamemerson Sadly, github doesn't make it entirely easy (that I know of) to see review histories, but the first answer is that /originally/ I wasn't using thread_local storage, so the utility was bit more obvious. The second answer is that the version without the user-specified engine is indeed probably not useful any more, since the internal engine is private anyway.  The only thing I'm thinking of at the moment would be a case where the user wanted their own syncronization against some other condition, but they could always do that externally as well.

This is suggesting to me that probably the right course of action is to make these implementation details, rather than part of the external interface.

adamemerson commented:
That sounds reasonable, then. Taking the mutex out of the interface. We can do thread-local with a user specified engine fairly easily, too.

chardan commented:
Next question: I'm close to having a patch for this, but do we prefer to remove all deprecated code or indicate this with a namespace? If the latter, is there an accepted way to indicate this in the unit test code? 
Alternatively, I could put it in another PR, unless we're 100% sure we want to jettison the code.

chardan commented:
Jenkins retest this please.

chardan commented:
Note: current commit removes the code.

chardan commented:
Jenkins retest this please.

adamemerson commented:
So! Sorry for being nitpicky about everything. Where do you imagine the per-thread RNGs being seeded? Since it only really needs to be done once per core, having random people do it whenever seems wasteful.

(Especially since if we're using a Mersenne Twister seeding and initializing takes a long time for things to rev up.)

I'd kind of recommend using something like the 'static optional' idiom I use in the gist.

Alternatively, we could do something analogous to what Seastar does and just initialize on engine per core in common_init. The downside is Ceph doesn't really HAVE any concept of cores and setting up some environment on each one. It might be nice if we DID but I don't know if we want to introduce it now.

chardan commented:
Jenkins retest this please.

chardan commented:
Jenkins retest this please.

chardan commented:
No worries! This is a much better time for scrutiny than six months from now 
when there's code using it. :-)

1) Right now, the mechanism if you're using the free functions is similar to calling
srand(). I like the Seastar idea best, the application layer feels like the place where
these decusions should be made.

2) static optional idiom is clever, but it ties everything to an implementation with
an extra runtime test for each number returned from the RNG. We could support this, but I feel
like it might be best accomplished through a policy class (basically, tool things to point
to compile-time selected function). 

3) The guts of randomize_rng() are in a detail namespace exactly for expanding on 
these ideas. Earlier in the review cycle, we removed explicit seeding functionality 
because it was considered dangerous. The only way to do it with an explicit value 
that remains is through the function object (reseedable) or function interfaces 
(seed once).

The short version I guess is that I feel like it's something best left to the application
as a general strategy, and WRT adopting a specific or custom behavior something that can
be done in a future PR.

chardan commented:
Aren't we adding lipstick to a pig? This is an improvement. Let's do it!

chardan commented:
Let's determine if we want to wind the world back to the version that I like best, which is one in which every use in Ceph I know of is modeled, and a number of future ones are also taken into account, or whether the current version is ok. I feel like the particulars of this are well-resolved, otherwise.

cbodley commented on src/include/random.h:
```diff
@@ -0,0 +1,389 @@
+
+template <int Min, int Max, int Seed,
+          typename EngineT = std::default_random_engine>
+auto make_random_number_function() -> int(*)()
```
these function interfaces seem very limited in utility because they only support `int`, and all of the min/max/seed values have to be known at compile time

in what cases would a user want to use `make_random_number_function()` over `random_number_generator`?

cbodley commented on src/test/common/test_random.cc:
```diff
@@ -0,0 +1,224 @@
+template <typename X0, typename X1, typename ...XS>
+bool none_equal(const X0 x0, const X1 x1, const XS ...xs)
+{
+  return none_equal(x0, none_equal(x1, xs...));
```
this doesn't look right? it compares `x0` with the boolean result of `none_equal()`?

cbodley commented on src/include/random.h:
```diff
@@ -0,0 +1,389 @@
+
+template <typename RealT, typename MutexT>
+RealT generate_random_number(const RealT max, MutexT& m,
+                             ceph::enable_if_t<std::is_floating_point<RealT>::value>* = nullptr)
```
there are still a few (unused?) overloads that mix a mutex argument with a thread-local engine

cbodley replied:
still unused

tchaikov commented on src/test/common/test_random.cc:
```diff
@@ -0,0 +1,224 @@
+  }
+ 
+  {
+    auto a = rng(10.0);
```
no, `rng` is an integer random number generator, and `a` will be an integer. this test is not verifying your expectation. instead, i think it demonstrates a misuse of this utility.

chardan commented:
@cbodley:
I) The remaining calls using the engine and mutex are
implementation details used by the class implementation, and also
are the entry point should we want to add any other behavior (the stuff
we just removed went through them also). They're in a detail namespace
and not part of the "public" interface, so they won't be accidentally
called.

II) I'm on the fence with the function version. The limitation of float is
not directly something we can get around, but the intended use of the
feature is that the RNG returned by make_random_function() is scoped and
independent, and its range and seed may be specified at compiler-time.    
Additionally, it may /not/ be reseeded, something desirable with the free
functions because of their shared engine. Besides these two differences,
it overlaps with the object-oriented interface. The intent is to occupy a
middle ground between those. The use-case difference probably isn't all
that strong, so it won't likely take much convincing to have me remove
them if that doesn't seem too compelling-- but, that's the "why" they're
there. :-)

III) Yeah, none_equal() is seriously dodgy-- it's really "pairwise ne",
and yep, boolean in the expansion-- I've changed it to something more
sensible that actually does what it says.

@tchaikov: You're correct, thanks for catching that! I've modified the
test code to improve clarity (and call the right RNG!).

chardan commented:
@cbodley I've gone ahead and taken the function-return version out, in their own commit. Easy enough to restore if desired.

How is everyone feeling?

chardan commented:
@cbodley I've hit on a way to check initialization that I think should work nicely! I've got to wrestle with cmake for a while on account of a testing issue, but I will check it in once I find a way to resolve that.

chardan commented:
@cbodley: I've added a way to check whether or not the free functions have been seeded. Note that the function object version doesn't need this since if you have an instance, it must have been seeded (there are no ctors that don't seed).

cbodley commented on src/include/random.h:
```diff
@@ -141,6 +153,11 @@ int generate_random_number(MutexT& m, EngineT& e)
+bool rng_is_seeded() 
+{
+  return detail::rng_is_seeded_state();
+}
```
i would really prefer automatic seeding, similar to what @adamemerson proposed. it seems like every user of this library would be forced into a pattern like this:
```c++
if (!rng_is_seeded()) {
  randomize_rng();
}
auto x = generate_random_number<int>(0, 10);
```

chardan replied:
That's correct, though hopefully at known points in the application. The issue I have with putting the test inside the generator is that you're checking a flag for each number you want, every time. This isn't going to be great in a tight loop where you want a lot of random numbers.

chardan replied:
Ok. Looking at the places we're using rand() right now, there don't seem to be many instances where my concern will be an actual issue-- we're typically just getting the occasional random number. I'll go ahead and implement "automagic".

cbodley replied:
> checking a flag for each number you want, every time

note that, at least for the thread-local interfaces, this check doesn't need to be atomic

adamemerson replied:
@chardan Also, mark the test as unlikely to hint to the branch predictor.

tchaikov commented on src/test/common/test_random.cc:
```diff
@@ -0,0 +1,205 @@
+template <typename X>
+void swallow_values(const X x)
+{
+ static_cast<void>(x);
```
indent.

tchaikov commented on src/test/common/test_random.cc:
```diff
@@ -0,0 +1,205 @@
+template <typename X, typename ...XS>
+void swallow_values(const X x, const XS... xs)
+{
+ swallow_values(x), swallow_values(xs...);
```
ditto

tchaikov commented on src/test/common/test_random.cc:
```diff
@@ -0,0 +1,205 @@
+
+  ceph::util::randomize_rng();          // change state of rng_is_seeded()
+
+  ASSERT_EQ(true, ceph::util::rng_is_seeded());   
```
ASSERT_TRUE

tchaikov commented on src/test/common/test_random.cc:
```diff
@@ -0,0 +1,205 @@
+TEST(util, test_random_canonical)
+{
+  // See if the random number generator has been seeded:
+  if(!ceph::util::rng_is_seeded()) {
```
add a space after `if`

cbodley commented on src/test/common/test_random.cc:
```diff
@@ -0,0 +1,205 @@
+
+/* This test needs to run before any other calls into the rng are made, because
+internally it uses thread_local storage; gtest runs these in one thread, so the
+expected false-state can't really be known. */
```
for tests that depend on tls, you should just wrap them in a lambda and run on a separate thread:
```c++
  std::thread([] {
    ASSERT_EQ(false, ceph::util::rng_is_seeded());
    ...
  }).join();
```
that way, we don't have to rely on test ordering, and arguments like `--gtest_repeat` will continue to work

chardan commented:
@cbodley @adamemerson @tchaikov @joscollin 

Ok-- after much discussion, the bottom line is that when we're calling rand() in our current code it's almost always for an "occasional" random number, not really things in tight loops. Therefore I've gone ahead and implemented "automagic" seeding. I think we're surely on the home stretch?

cbodley commented on src/include/random.h:
```diff
@@ -0,0 +1,354 @@
+
+template <typename MutexT, 
+          typename EngineT = std::default_random_engine>
+void randomize_rng(MutexT& m)
```
unused

cbodley commented on src/include/random.h:
```diff
@@ -0,0 +1,354 @@
+
+template <typename MutexT, 
+          typename EngineT = std::default_random_engine>
+void randomize_rng(const int n, MutexT& m)
```
unused

cbodley commented on src/include/random.h:
```diff
@@ -0,0 +1,354 @@
+          typename DistributionT,
+          typename EngineT>
+NumberT generate_random_number(const NumberT min, const NumberT max,
+                               MutexT& m)
```
unused

cbodley commented on src/include/random.h:
```diff
@@ -0,0 +1,354 @@
+
+template <typename IntegerT, typename MutexT>
+IntegerT generate_random_number(const IntegerT min, const IntegerT max, MutexT& m,
+                                ceph::enable_if_t<std::is_integral<IntegerT>::value>* = nullptr)
```
unused

cbodley commented on src/include/random.h:
```diff
@@ -0,0 +1,354 @@
+template <typename IntegerT, typename MutexT>
+int generate_random_number(const IntegerT max,
+                           MutexT& m,
+                           ceph::enable_if_t<std::is_integral<IntegerT>::value>* = nullptr)
```
unused

cbodley commented on src/include/random.h:
```diff
@@ -0,0 +1,354 @@
+
+template <typename RealT, typename MutexT>
+RealT generate_random_number(const RealT min, const RealT max, MutexT& m,
+                             ceph::enable_if_t<std::is_floating_point<RealT>::value>* = nullptr)
```
unused

cbodley commented:
> when we're calling rand() in our current code it's almost always for an "occasional" random number, not really things in tight loops. Therefore I've gone ahead and implemented "automagic" seeding.

the new seeding looks good, thanks! since it only costs a single branch, i'm not really worried about performance in a tight loop. if there are special cases where this actually becomes measurable, we can use the `<random>` interfaces directly

> I think we're surely on the home stretch?

agreed, i think we've narrowed the scope to a point where it's hard to misuse, and fast/easy to use correctly

as a future enhancement, we might consider specializing the seeding logic (as @adamemerson proposed with [hardware_seed_seq](https://gist.github.com/adamemerson/ca2a8fa70f2fc592740469372ee83026#file-random-h-L47) above) for engines such as `mt19937` that can't be seeded properly with a single int

chardan commented:
Jenkins retest this please.

chardan commented:
@cbodley Agreed. Please have a peek, things should be attended to (exception, #253 is an active codepath).

Assuming there are no further comments, thanks to everyone for your energy and input on this! I think it will accomplish its goals and allow us to move away from rand()!

tchaikov commented on src/include/random.h:
```diff
@@ -259,7 +193,7 @@ RealT generate_random_number(const RealT min, const RealT max,
                                         std::default_random_engine>
                                        (min, max);
-}
+} 
```
trailing space

tchaikov commented:
need to squash the fix commits into the ones they are fixing.

chardan commented:
@tchaikov Ok with this squishing, or do you want all of the review deltas just merged into a single change? I'm ok with things as they are, but no strong feelings one way or another.

chardan commented:
@liewegas Also (once more, in my best Rod Serling impression) "Submitted for your approval."

tchaikov commented on src/include/random.h:
```diff
@@ -0,0 +1,288 @@
+{
+  thread_local boost::optional<EngineT> rng_engine;
+
+  if(!rng_engine) {
```
a space after `if`.

chardan commented:
@tchaikov Ok, revised! Thank you.

chardan commented:
@tchaikov @cbodley @joscollin @adamemerson 
nudge!

tchaikov commented:
as long as shaman is happy

liewegas commented:
http://pulpito.ceph.com/sage-2017-09-01_21:29:08-rados-wip-sage-testing-2017-09-01-1033-distro-basic-smithi/

chardan commented:
Yay! Thanks, everyone! Some work on actually replacing rand() calls will be coming soon to a PR near you!
